### PR TITLE
Adds driver for Micro-Epsilon ILD1900 sensor

### PR DESCRIPTION
**Pull request by d-loret**
Summary:
Adds the JSD driver for the Micro-Epsilon optoNCDT ILD1900 laser
sensor which measures distance with high precision.

Test Plan:
Performed preliminary testing via the written test executable and
the terminal connected to a local EtherCAT network. No inputs were
connected yet.

Reviewers: alex-brinkman, kwehage

### CODE REVIEW

File: src/jsd_ild1900.c [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+  while (n) {
+    num_bits += (n & 1);
+    n >>= 1;
+  }
+  if (num_bits == 1) {
+    return true;
+  } else {
+    return false;
+  }
+}
```
Comment (alex-brinkman) [inline src/jsd_ild1900.c:104]: nice!

File: src/jsd_ild1900_types.h [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  double measuring_rate;  ///< Number of measurements per second expressed in
+                          ///< Hz. Maximum value is 10000 Hz. See Operating
+                          ///< Instructions optoNCDT 1900-IE EtherCAT
+                          ///< section 7.4.3.
+  uint32_t
+      averaging_number;  ///< Number of consecutive measurements averaged
+                         ///< together. If averaging_type is
+                         ///< JSD_ILD1900_AVERAGING_NONE, this value is ignored.
+  jsd_ild1900_averaging_t
+                               averaging_type;  ///< Type of averaging formula applied to measurements
```
Comment (alex-brinkman) [inline src/jsd_ild1900_types.h:96]: nitpick: I find this whitespacing odd but it's immaterial to me

File: src/jsd_ild1900_types.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  JSD_ILD1900_ERROR_LASER_OFF                     ///< Laser is off
+} jsd_ild1900_error_t;
+
+/**
+ * @brief Read struct for ILD1900 device
+ */
+typedef struct {
+  double distance;  ///< Distance from the sensor to the target in m
+  double
+           unlinearized_center_of_gravity;  ///< Unlinearized distance in percentage
```
Comment (alex-brinkman) [inline src/jsd_ild1900_types.h:122]: nit whitespacing

File: src/jsd_ild1900_types.h [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  double   intensity;      ///< Signal intensity of peak in percentage
+  uint32_t timestamp;      ///< Timestamp of measurement in nanoseconds
+  uint32_t
+           counter;  ///< Measurement counter. Distinguishes separate measurements.
```
Comment (alex-brinkman) [inline src/jsd_ild1900_types.h:126]: nit whitespacing

File: src/jsd_ild1900.c [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  state->timestamp               = txpdo->timestamp;
+  state->counter                 = txpdo->counter;
+  state->sensor_status           = txpdo->sensor_status;
+  state->linearized_distance_raw = txpdo->linearized_distance_raw;
+  state->intensity               = (100 * txpdo->intensity_raw) / 1023.0;
+  state->unlinearized_center_of_gravity =
+      (100 * txpdo->unlinearized_distance_raw) / 262143.0;
+  state->distance = ((int32_t)txpdo->linearized_distance_raw - 98232) /
+                        65536.0 * JSD_ILD1900_MR_MAP[config->model] +
+                    JSD_ILD1900_SMR_MAP[config->model];
```
Comment (alex-brinkman) [inline src/jsd_ild1900.c:50]: image
I believe you are mixing units of mm with meters in this calc, please review.

File: src/jsd_ild1900.c [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+  /*
+   * TxPDO Mapping
+   * 0x1A04 - Timestamp (0x6002:1)
+   * 0x1A08 - Measurement counter (0x6003:1)
+   * 0x1A0C - Status (0x6004:1)
+   * 0x1A10 - Not linearized distance + intensity + distance (0x6005:1 +
+   * 0x5006:1 + 0x6007:1)
+   *
+   * ILD1900 sensors do not support Complete Access. Steps to set TxPDO
```
Comment (alex-brinkman) [inline src/jsd_ild1900.c:168]: Sorry CA didn't pan out for this sensor, good job tracking that down!

**Review by alex-brinkman — CHANGES_REQUESTED** [review]
Great Job Daniel, You did an excellent job documenting the code and respecting the intended private/public API exposure!

File: src/jsd_ild1900.c (hunk 1 above)
Comment (alex-brinkman) [reply src/jsd_ild1900.c:104]: I was referring to the powerof2() function

File: src/jsd_ild1900_types.h (hunk 2 above)
Comment (d-loret) [reply src/jsd_ild1900_types.h:96]: It's git clang-format. For some reason it decides to organize it like that.

File: src/jsd_ild1900.c (hunk 5 above)
Comment (d-loret) [reply src/jsd_ild1900.c:51]: Discussed offline. In the code all units are in meters, including MR.

**Comment by d-loret** [comment]
@alex-brinkman, I updated based on our findings when we tested in the lab. I am also excluding linearized and unlinearized distances from the state for simplicity's sake (I don't think we can get very helpful information from them).

**Pull request merged**